README.md
# NegativeScreen #

https://zerowidthjoiner.net/negativescreen

***

## Description

NegativeScreen's main goal is to support your poor tearful eyes when enjoying the bright white interweb in a dark room.
This task is joyfully achieved by inverting the colors of your screen.

Unlike the Windows Magnifier, which is also capable of such color inversion,
NegativeScreen was specifically designed to be easy and convenient to use.

It comes with a minimal graphic interface in the form of a system tray icon with a context menu,
but don't worry, this only makes it easier to use!


## Features

Invert screen's colors.

Additionally, many color effects can be applied.
For example, different inversion modes, including "smart" modes,
swapping blacks and whites while keeping colors (about) the sames.

You can now configure the color effects manually via a configuration file.
You can also configure the hot keys for every actions, using the same configuration file.

A basic web api is part of NegativeScreen >= 2.5
It is disabled by default. When enabled, it listens by default on port 8990, localhost only.
See the configuration file to enable the api or change the listening uri...

All commands must be sent with the POST http method.
The following commands are implemented:
- TOGGLE
- ENABLE
- DISABLE
- SET "Color effect name" (without the quotes)

Any request sent with a method other than POST will not be interpreted,
and a response containing the application version will be sent.


## Requirements

NegativeScreen < 2.0 needs at least Windows Vista to run.

Versions 2.0+ need at least Windows 7.

Both run on Windows 8 or superior.

Graphic acceleration (Aero) must be enabled.

.NET 4.5 is required. (Should be available out of the box on up-to-date Windows installations)


## Default controls

-Press Win+Alt+H to Halt the program immediately
-Press Win+Alt+N to toggle color inversion (mnemonic: Night vision :))

-Press Win+Alt+F1-to-F10 to switch between inversion modes:
	* F1: standard inversion
	* F2: smart inversion1 - theoretical optimal transformation (but ugly desaturated pure colors)
	* F3: smart inversion2 - high saturation, good pure colors
	* F4: smart inversion3 - overall desaturated, yellows and blues plain bad. actually relaxing and very usable
	* F5: smart inversion4 - high saturation. yellows and blues  plain bad. actually quite readable
	* F6: smart inversion5 - not so readable. good colors. (CMY colors a bit desaturated, still more saturated than normal)
	* F7: negative sepia
	* F8: negative gray scale
	* F9: negative red
	* F10: red
	* F11: grayscale


## Configuration file

A customizable configuration file is created the first time you use "Edit Configuration" from the context menu.

The default location for this file is next to NegativeScreen.exe, and is called "negativescreen.conf"

If the default location is inaccessible,
NegativeScreen will try to create the configuration file in %AppData%/NegativeScreen/negativescreen.conf

This feature allows to deploy NegativeScreen.exe in a read-only location for unprivileged users.
Each user can then have its own configuration file.

The order of priority for trying to read a configuration file when starting NegativeScreen is as follows:
- %AppData%/NegativeScreen/negativescreen.conf
- negativescreen.conf in the directory where NegativeScreen.exe is located
- If the above fails, the embedded default configuration is used

Should something go wrong (syntax error, bad hot key...), you can simply delete the configuration file,
the internal default configuration will be used.

If the configuration file is missing, you can use the "Edit Configuration" menu to regenerate the default one.

Syntax: see in the configuration file...


***

Many thanks to Tom MacLeod who gave me the idea for the "smart" inversion mode :)


Enjoy!


CHANGELOG.txt
Version 2.6.1, 2021-04-10
------------------------
- [changed] target .NET 4.5 (see https://zerowidthjoiner.net/negativescreen#comment-534224)

Version 2.6, 2018-12-05
------------------------
- [added] new matrices (color blindness simulation, and binary) added to the default configuration file
- [fixed] color effect sometimes randomly activated by other running programs on the system

Version 2.5, 2017-05-28
------------------------
- [fixed] when an instance is already running, enable its color effect instead of showing an unhelpful message about it
- [fixed] no longer crash when trying to set a color effect while the Windows Magnifier color inversion is enabled
- [added] a basic web api is now part of NegativeScreen! See the readme for more information
- [added] new EnableApi configuration option, defaults to false
- [added] new ApiListeningUri configuration option, defaults to listening on port 8990, localhost only
- [added] new About window, available from the systray menu
- [added] add a retro-compatible user-specific fallback location for the configuration file:
		- if the file "%AppData%/NegativeScreen/negativescreen.conf" exists, it will be used in priority.
		- if the "negativescreen.conf" file beside the executable cannot be modified,
			trying to edit the current configuration will result in a configuration in AppData being created.

Version 2.4, 2014-11-30
------------------------
- [fixed] force the working directory to be the one of the executable (so the .conf file is found when starting from the cli)

Version 2.3, 2014-03-28
------------------------
- [fixed] crash if more than one custom color effect is added without an associated hot key
- [added] (re)added the ShowAeroWarning configuration option from the version 1.*
- [added] check box next to the current color effect in the right click menu

Version 2.2, 2013-02-17
------------------------
- [fixed] avoid crash if a hot key cannot be registered
- [added] new ActiveOnStartup configuration option

Version 2.1, 2012-12-22
------------------------
- [fixed] crash on Windows 8 due to inter-threads calls
		(this was not a problem on Windows 7)
- [added] new MainLoopRefreshTime configuration option

Version 2.0, 2012-12-14
------------------------
Major update:
	- reworked entirely
	- better performances
	- smooth transitions
	- graphic interface (minimal)
	- configuration file:
		- fine grained configuration
		- custom hotkeys
		- custom color effects

Version 1.14, 2014-11-30
------------------------
- [fixed] force the working directory to be the one of the executable (so the .conf file is found when starting from the cli)

Version 1.13, 2013-07-04
------------------------
- [added] implemented the configuration file for the version 1.*
- [added] configuration option to disable the message warning about aero being disabled

Version 1.12, 2012-12-14
------------------------
- [fixed] crash on Windows 8

Version 1.11, 2012-06-26
------------------------
- [added] new red and negative red filters bound on F9 and F10

Version 1.10, 2012-05-15
------------------------
- [fixed] infinite flickering on Vista (introduced in 1.8)

Version 1.9, 2012-04-16
------------------------
- [fixed] stupid bug if the Windows taskbar is vertical

Version 1.8, 2012-03-22
------------------------
- [fixed] execution is no longer prevented if aero is disabled (previous bug corrected)
		however, the performances will still be horrible

Version 1.7, 2012-03-21
------------------------
- [fixed] works as expected when using custom DPI settings

Version 1.6, 2012-02-27
------------------------
- [fixed] crash on X86 on launch
- [fixed] prevent execution if aero is disabled
		(prevent undesirable behaviours: black screens, 100% CPU usage...)

Version 1.5, 2012-02-16
------------------------
- [fixed] (internal) ColorMatrix implementation
- [added] new feature: choice between 9 inversion modes
		(smart mode, etc... see readme for details)

Version 1.4, 2012-02-09
------------------------
- [fixed] does not crash anymore on Windows Vista

Version 1.3, 2012-02-01
------------------------
- [fixed] multi-screen should finally work!
		after a lot of tests and coding, I re-thought completely the architecture
- [knownbug] in some screen configurations, if the primary screen if smaller than the other screen,
		 the second larger screen has a black border at its bottom.
		 It could be a bug in the Windows API...

Version 1.2, 2011-09-18
------------------------
- [fixed] multi-screen support (again) : bug with main screen on the right

Version 1.1, 2011-09-13
------------------------
- [fixed] multi-screen support
- [fixed] when halted while paused, the application never stopped

Version 1.0, 2011-09-06
------------------------
- Initial release

NegativeScreen/Configuration.cs
// Copyright 2011-2017 Melvyn Laïly
// https://zerowidthjoiner.net

// This file is part of NegativeScreen.

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

using System;
using System.Collections.Generic;
using System.Text;
using System.Windows.Forms;
using System.Linq;
using System.IO;
using System.Diagnostics;
using System.Security.Permissions;

namespace NegativeScreen
{
	class Configuration : IConfigurable
	{
		public string WorkingDirectoryConfigurationFileName { get; } =
			Path.Combine(Environment.CurrentDirectory, "negativescreen.conf");
		public string AppDataConfigurationFileName { get; } =
			Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "NegativeScreen/negativescreen.conf");
		public ConfigurationLocation Source { get; private set; }

		#region Default configuration
		public const string DefaultConfiguration =
@"# comments: if the character '#' is found, the rest of the line is ignored.
# quotes: allow to place a '#' inside a value. they do not appear in the final result.
# i.e. blah=""hello #1!"" will create a parameter blah with a value of: hello #1!
# To place a quotation mark inside quotes, double it.
# i.e. blah=""hello"""""" will create a parameter blah with a value of: hello""

#Predefined keys
# You can use the following modifiers: alt, ctrl, shift, win
# and a key from http://msdn.microsoft.com/en-us/library/system.windows.forms.keys%28v=vs.71%29.aspx
# You can either use its textual representation, or its numerical value.
# WARNING: if the key is not valid, the program will probably crash...

Toggle=win+alt+N
Exit=win+alt+H

SmoothTransitions=true
SmoothToggles=true

# in miliseconds
MainLoopRefreshTime=100

InitialColorEffect=""Smart Inversion""

ActiveOnStartup=true

ShowAeroWarning=true

EnableApi=false
# A URI prefix string is composed of a scheme (http), a host, an optional port, and an optional path. The trailing slash is mandatory.
# To specify that the HttpListener accepts all requests sent to a port, replace the host element with the ""+"" character: ""https://+:8080"".
# More information on listening prefixes can be found here: https://msdn.microsoft.com/en-us/library/system.net.httplistener%28v=vs.110%29.aspx
# Please note setting a host other than localhost will require admin privileges.
# If you do so, also remember to add an exception to your firewall.
ApiListeningUri=http://localhost:8990/

# Matrices definitions
# The left hand is used as a description, while the right hand is broken down in two parts:
# - the hot key combination, followed by a new line, (this part is optional)
# - the matrix definition, with or without new lines between rows.
# The matrices must have 5 rows and 5 columns,
# each line between curved brackets,
# the elements separated by commas.
# The decimal separator is a dot.

# I'm not too keen on the maths, so maybe I'm missing some things,
# but here is my basic understanding of what does what in a color matrix, when applied to a color vector:
# r*=x    g+=x*r  b+=x*r  a+=x*r  0
# r+=x*g  g*=x    b+=x*g  a+=x*g  0
# r+=x*b  g+=x*b  b*=x    a+=x*b  0
# r+=x*a  g+=x*a  b+=x*a  a*=x    0
# r+=x    g+=x    b+=x    a+=x    1

# where x is the value in the matrix.

Simple Inversion=win+alt+F1
{ -1,  0,  0,  0,  0 }
{  0, -1,  0,  0,  0 }
{  0,  0, -1,  0,  0 }
{  0,  0,  0,  1,  0 }
{  1,  1,  1,  0,  1 }

# Theoretical optimal transfomation (but ugly desaturated pure colors due to ""overflows""...)
# Many thanks to Tom MacLeod who gave me the idea for these inversion modes.
Smart Inversion=win+alt+F2
{  0.3333333, -0.6666667, -0.6666667,  0.0000000,  0.0000000 }
{ -0.6666667,  0.3333333, -0.6666667,  0.0000000,  0.0000000 }
{ -0.6666667, -0.6666667,  0.3333333,  0.0000000,  0.0000000 }
{  0.0000000,  0.0000000,  0.0000000,  1.0000000,  0.0000000 }
{  1.0000000,  1.0000000,  1.0000000,  0.0000000,  1.0000000 }

# High saturation, good pure colors.
Smart Inversion Alt 1=win+alt+F3
{  1, -1, -1,  0,  0 }
{ -1,  1, -1,  0,  0 }
{ -1, -1,  1,  0,  0 }
{  0,  0,  0,  1,  0 }
{  1,  1,  1,  0,  1 }

# Overall desaturated, yellows and blue plain bad. actually relaxing and very usable.
Smart Inversion Alt 2=win+alt+F4
{  0.39, -0.62, -0.62,  0.00,  0.00 }
{ -1.21, -0.22, -1.22,  0.00,  0.00 }
{ -0.16, -0.16,  0.84,  0.00,  0.00 }
{  0.00,  0.00,  0.00,  1.00,  0.00 }
{  1.00,  1.00,  1.00,  0.00,  1.00 }

# High saturation. yellows and blues plain bad. actually quite readable.
Smart Inversion Alt 3=win+alt+F5
{  1.0895080, -0.9326327, -0.9326330,  0.0000000,  0.0000000 }
{ -1.8177180,  0.1683074, -1.8416920,  0.0000000,  0.0000000 }
{ -0.2445895, -0.2478156,  1.7621850,  0.0000000,  0.0000000 }
{  0.0000000,  0.0000000,  0.0000000,  1.0000000,  0.0000000 }
{  1.0000000,  1.0000000,  1.0000000,  0.0000000,  1.0000000 }

# Not so readable, good colors (CMY colors a bit desaturated, still more saturated than normal).
Smart Inversion Alt 4=win+alt+F6
{  0.50, -0.78, -0.78,  0.00,  0.00 }
{ -0.56,  0.72, -0.56,  0.00,  0.00 }
{ -0.94, -0.94,  0.34,  0.00,  0.00 }
{  0.00,  0.00,  0.00,  1.00,  0.00 }
{  1.00,  1.00,  1.00,  0.00,  1.00 }

Negative Sepia=win+alt+F7
{ -0.393, -0.349, -0.272,  0.000,  0.000 }
{ -0.769, -0.686, -0.534,  0.000,  0.000 }
{ -0.189, -0.168, -0.131,  0.000,  0.000 }
{  0.000,  0.000,  0.000,  1.000,  0.000 }
{  1.351,  1.203,  0.937,  0.000,  1.000 }

Negative Grayscale=win+alt+F8
{ -0.3, -0.3, -0.3,  0.0,  0.0 }
{ -0.6, -0.6, -0.6,  0.0,  0.0 }
{ -0.1, -0.1, -0.1,  0.0,  0.0 }
{  0.0,  0.0,  0.0,  1.0,  0.0 }
{  1.0,  1.0,  1.0,  0.0,  1.0 }

#Grayscaled
Negative Red=win+alt+F9
{ -0.3,  0.0,  0.0,  0.0,  0.0 }
{ -0.6,  0.0,  0.0,  0.0,  0.0 }
{ -0.1,  0.0,  0.0,  0.0,  0.0 }
{  0.0,  0.0,  0.0,  1.0,  0.0 }
{  1.0,  0.0,  0.0,  0.0,  1.0 }

#Grayscaled
Red=win+alt+F10
{  0.3,  0.0,  0.0,  0.0,  0.0 }
{  0.6,  0.0,  0.0,  0.0,  0.0 }
{  0.1,  0.0,  0.0,  0.0,  0.0 }
{  0.0,  0.0,  0.0,  1.0,  0.0 }
{  0.0,  0.0,  0.0,  0.0,  1.0 }

Grayscale=win+alt+F11
{ 0.3,  0.3,  0.3,  0.0,  0.0 }
{ 0.6,  0.6,  0.6,  0.0,  0.0 }
{ 0.1,  0.1,  0.1,  0.0,  0.0 }
{ 0.0,  0.0,  0.0,  1.0,  0.0 }
{ 0.0,  0.0,  0.0,  0.0,  1.0 }

# Color blindness simulation matrices
# Source: http://web.archive.org/web/20081014161121/http://www.colorjack.com/labs/colormatrix/

# http://www.color-blindness.com/protanopia-red-green-color-blindness/
# Red-Green Color Blindness   - Male Population: 1.01%, Female 0.02
Color blindness simulation: Protanopia (Red-Green Color Blindness)=
{ 0.567, 0.558, 0, 0, 0 }
{ 0.433, 0.442, 0.242, 0, 0 }
{ 0, 0, 0.758, 0, 0 }
{ 0, 0, 0, 1, 0 }
{ 0, 0, 0, 0, 1 }


# http://www.color-blindness.com/protanopia-red-green-color-blindness/
# Protanomaly (red-weak)  - Male Population: 1.08%, 0.03%
Color blindness simulation: Protanomaly (red-weak)=
{ 0.817, 0.333, 0, 0, 0 }
{ 0.183, 0.667, 0.125, 0, 0 }
{ 0, 0, 0.875, 0, 0 }
{ 0, 0, 0, 1, 0 }
{ 0, 0, 0, 0, 1 }


# http://www.color-blindness.com/deuteranopia-red-green-color-blindness/
# http://www.colour-blindness.com/general/prevalence/
# Deuteranopia (also called green-blind) - Male Population: 1%, Female Population: 0.1%
Color blindness simulation: Deuteranopia (green-blind)=
{ 0.625, 0.7, 0, 0, 0 }
{ 0.375, 0.3, 0.3, 0, 0 }
{ 0, 0, 0.7, 0, 0 }
{ 0, 0, 0, 1, 0 }
{ 0, 0, 0, 0, 1 }


# http://www.colour-blindness.com/general/prevalence/
# Deuteranomaly (green-weak) - Male Population: 5%, Female Population: 0.35%
Color blindness simulation: Deuteranomaly (green-weak)=
{ 0.8, 0.258, 0, 0, 0 }
{ 0.2, 0.742, 0.142, 0, 0 }
{ 0, 0, 0.858, 0, 0 }
{ 0, 0, 0, 1, 0 }
{ 0, 0, 0, 0, 1 }


# http://www.color-blindness.com/tritanopia-blue-yellow-color-blindness/
# http://www.colour-blindness.com/general/prevalence/
# Tritanopia – Blue-Yellow Color Blindness - rare. Some sources estimate 0.008% 
Color blindness simulation: Tritanopia (Blue-Yellow Color Blindness - rare)=
{ 0.95, 0, 0, 0, 0 }
{ 0.05, 0.433, 0.475, 0, 0 }
{ 0, 0.567, 0.525, 0, 0 }
{ 0, 0, 0, 1, 0 }
{ 0, 0, 0, 0, 1 }


# http://www.color-blindness.com/tritanopia-blue-yellow-color-blindness/
# http://www.colour-blindness.com/general/prevalence/
# Tritanomaly (blue-yellow weak) - Male 0.01%, Female 0.01%
Color blindness simulation: Tritanomaly (blue-yellow weak)=
{ 0.967, 0, 0, 0, 0 }
{ 0.033, 0.733, 0.183, 0, 0 }
{ 0, 0.267, 0.817, 0, 0 }
{ 0, 0, 0, 1, 0 }
{ 0, 0, 0, 0, 1 }


# http://www.color-blindness.com/2007/07/20/monochromacy-complete-color-blindness/
# Total color blindness - Occurrences are estimated to be between 1 : 30’000 and 1 : 50’000.
Color blindness simulation: Achromatopsia (Total color blindness)=
{ 0.299, 0.299, 0.299, 0, 0 }
{ 0.587, 0.587, 0.587, 0, 0 }
{ 0.114, 0.114, 0.114, 0, 0 }
{ 0, 0, 0, 1, 0 }
{ 0, 0, 0, 0, 1 }


# http://www.color-blindness.com/2007/07/20/monochromacy-complete-color-blindness/
# All color-weak - Different sources vary between 1 in 33’000 to 100’000 (0.001%).
Color blindness simulation: Achromatomaly (Total color weakness)=
{ 0.618, 0.163, 0.163, 0, 0 }
{ 0.32, 0.775, 0.32, 0, 0 }
{ 0.062, 0.062, 0.516, 0, 0 }
{ 0, 0, 0, 1, 0 }
{ 0, 0, 0, 0, 1 }

Binary (Black and white)=
{  127,  127,  127,  0,  0 }
{  127,  127,  127,  0,  0 }
{  127,  127,  127,  0,  0 }
{  0,  0,  0,  1,  0 }
{ -180, -180, -180,  0,  1 }
";
		#endregion

		private static Configuration _current;
		public static Configuration Current
		{
			get
			{
				Initialize();
				return _current;
			}
		}

		public static void Initialize()
		{
			if (_current == null)
			{
				_current = new Configuration();
			}
		}

		private Configuration()
		{
			ColorEffects = new List<KeyValuePair<HotKey, ScreenColorEffect>>();

			string configFileContent = ReadCurrentConfiguration();

			Parser.AssignConfiguration(configFileContent, this, new HotKeyParser(), new MatrixParser());
			if (!string.IsNullOrWhiteSpace(InitialColorEffectName))
			{
				try
				{
					this.InitialColorEffect = this.ColorEffects.Single(x =>
						x.Value.Description.ToLowerInvariant() == InitialColorEffectName.ToLowerInvariant()).Value;
				}
				catch (Exception)
				{
					// probably not ideal
					this.InitialColorEffect = new ScreenColorEffect(BuiltinMatrices.Negative, "Negative");
				}
			}
		}

		private string ReadCurrentConfiguration()
		{
			try
			{
				// First try to read the AppData conf
				Source = ConfigurationLocation.AppData;
				return File.ReadAllText(AppDataConfigurationFileName);
			}
			catch (Exception)
			{
				try
				{
					// If we can't access it, try the one in the working directory
					Source = ConfigurationLocation.WorkingDirectory;
					return File.ReadAllText(WorkingDirectoryConfigurationFileName);
				}
				catch (Exception)
				{
					// If all else fails, read the default embedded one
					Source = ConfigurationLocation.Embedded;
					return DefaultConfiguration;
				}
			}
		}

		private void EnsureAppDataConfigurationFileExists()
		{
			Directory.CreateDirectory(Path.GetDirectoryName(AppDataConfigurationFileName));
			if (!File.Exists(AppDataConfigurationFileName))
			{
				File.WriteAllText(AppDataConfigurationFileName, DefaultConfiguration);
			}
		}

		public static void UserEditCurrentConfiguration()
		{
			void EditPath(string path)
			{
				Process.Start("notepad", path);
			}

			switch (Current.Source)
			{
				case ConfigurationLocation.AppData:
					// If the source is AppData, use it:
					EditPath(Current.AppDataConfigurationFileName);
					break;
				case ConfigurationLocation.WorkingDirectory:
				case ConfigurationLocation.Embedded:
				default:
					// If the source is the working directory, or no config file exists yet, try there first:
					// (File.GetAccessControl() looks very scary and error prone, so we just try to open the file...)
					try
					{
						if (!File.Exists(Current.WorkingDirectoryConfigurationFileName))
						{
							// create it with the default configuration template
							File.WriteAllText(Current.WorkingDirectoryConfigurationFileName, DefaultConfiguration);
						}
						else
						{
							// just check we can access it
							using (new FileStream(Current.WorkingDirectoryConfigurationFileName, FileMode.Open, FileAccess.ReadWrite)) { }
						}
						// Edit the working directory conf if we can
						EditPath(Current.WorkingDirectoryConfigurationFileName);
					}
					catch (Exception)
					{
						// If we can't access the working directory conf, use the AppData one instead
						Current.EnsureAppDataConfigurationFileExists();
						EditPath(Current.AppDataConfigurationFileName);
					}
					break;
			}
		}

		[MatchingKey("Toggle", CustomParameter = HotKey.ToggleKeyId)]
		public HotKey ToggleKey { get; protected set; }

		[MatchingKey("Exit", CustomParameter = HotKey.ExitKeyId)]
		public HotKey ExitKey { get; protected set; }

		[MatchingKey("SmoothTransitions")]
		public bool SmoothTransitions { get; protected set; }

		[MatchingKey("SmoothToggles")]
		public bool SmoothToggles { get; protected set; }

		[MatchingKey("MainLoopRefreshTime", CustomParameter = 100)]
		public int MainLoopRefreshTime { get; protected set; }

		[MatchingKey("ActiveOnStartup", CustomParameter = true)]
		public bool ActiveOnStartup { get; protected set; }

		[MatchingKey("ShowAeroWarning", CustomParameter = true)]
		public bool ShowAeroWarning { get; protected set; }

		[MatchingKey("EnableApi", CustomParameter = false)]
		public bool EnableApi { get; protected set; }

		[MatchingKey("ApiListeningUri", CustomParameter = "http://localhost:8990/")]
		public string ApiListeningUri { get; protected set; }

		[MatchingKey("InitialColorEffect")]
		public string InitialColorEffectName { get; protected set; }

		public ScreenColorEffect InitialColorEffect { get; protected set; }

		public List<KeyValuePair<HotKey, ScreenColorEffect>> ColorEffects { get; protected set; }

		public void HandleDynamicKey(string key, string value)
		{
			// value is already trimmed
			if (value.StartsWith("{"))
			{
				// no hotkey
				this.ColorEffects.Add(new KeyValuePair<HotKey, ScreenColorEffect>(
					HotKey.Empty,
					new ScreenColorEffect(MatrixParser.StaticParseMatrix(value), key)));
			}
			else
			{
				// first part is the hotkey, second part is the matrix
				var splitted = value.Split(new char[] { '\n' }, 2);
				if (splitted.Length < 2)
				{
					throw new Exception(string.Format(
						"The value assigned to \"{0}\" is unexpected. The hotkey must be separated from the matrix by a new line.",
						key));
				}
				this.ColorEffects.Add(new KeyValuePair<HotKey, ScreenColorEffect>(
					HotKeyParser.StaticParse(splitted[0]),
					new ScreenColorEffect(MatrixParser.StaticParseMatrix(splitted[1]), key)));
			}
		}
	}

	class HotKeyParser : ICustomParser
	{

		public Type ReturnType => typeof(HotKey);

		public object Parse(string rawValue, object customParameter)
		{
			int defaultId = -1;
			if (customParameter is int)
			{
				defaultId = (int)customParameter;
			}
			return StaticParse(rawValue, defaultId);
		}

		public static HotKey StaticParse(string rawValue, int defaultId = -1)
		{
			KeyModifiers modifiers = KeyModifiers.NONE;
			Keys key = Keys.None;
			string trimmed = rawValue.Trim();
			var splitted = trimmed.Split('+');
			foreach (var item in splitted)
			{
				// modifier
				switch (item.ToLowerInvariant())
				{
					case "alt":
						modifiers |= KeyModifiers.MOD_ALT;
						break;
					case "ctrl":
						modifiers |= KeyModifiers.MOD_CONTROL;
						break;
					case "shift":
						modifiers |= KeyModifiers.MOD_SHIFT;
						break;
					case "win":
						modifiers |= KeyModifiers.MOD_WIN;
						break;
					default:
						// key
						if (!Enum.TryParse(item, true, out key))
						{
							// try to parse numeric value
							if (int.TryParse(item, out int numericValue))
							{
								if (Enum.IsDefined(typeof(Keys), numericValue))
								{
									key = (Keys)numericValue;
								}
							}
						}
						break;
				}

			}
			return new HotKey(modifiers, key, defaultId);
		}
	}

	class MatrixParser : ICustomParser
	{

		public Type ReturnType => typeof(float[,]);

		public object Parse(string rawValue, object customParameter)
			=> StaticParseMatrix(rawValue);

		public static float[,] StaticParseMatrix(string rawValue)
		{
			float[,] matrix = new float[5, 5];
			var rows = System.Text.RegularExpressions.Regex.Matches(rawValue, @"{(?<row>.*?)}",
				System.Text.RegularExpressions.RegexOptions.ExplicitCapture);
			if (rows.Count != 5)
			{
				throw new Exception("The matrices must have 5 rows.");
			}
			for (int x = 0; x < rows.Count; x++)
			{
				var row = rows[x];
				var columnSplit = row.Groups["row"].Value.Split(',');
				if (columnSplit.Length != 5)
				{
					throw new Exception("The matrices must have 5 columns.");
				}
				for (int y = 0; y < matrix.GetLength(1); y++)
				{
					if (!float.TryParse(columnSplit[y],
						System.Globalization.NumberStyles.Float,
						System.Globalization.NumberFormatInfo.InvariantInfo,
						out float value))
					{
						throw new Exception(string.Format("Unable to parse \"{0}\" to a float.", columnSplit[y]));
					}
					matrix[x, y] = value;
				}
			}
			return matrix;
		}
	}

	struct HotKey
	{
		public const int ToggleKeyId = 42;
		public const int ExitKeyId = 43;

		private static int CurrentId = 100;

		public static readonly HotKey Empty = new HotKey()
		{
			Id = 0,
			Key = Keys.None,
			Modifiers = KeyModifiers.NONE
		};

		public KeyModifiers Modifiers { get; private set; }
		public Keys Key { get; private set; }
		public int Id { get; private set; }

		public HotKey(KeyModifiers modifiers, Keys key, int id = -1)
			: this()
		{
			Modifiers = modifiers;
			// 65535
			Key = key & Keys.KeyCode;
			if (id == -1)
			{
				Id = CurrentId;
				CurrentId++;
			}
			else
			{
				Id = id;
			}
		}

		public override int GetHashCode()
			=> (int)Key | (int)Modifiers << 16 | Id << 20;

		public override bool Equals(object obj)
		{
			if (obj == null)
			{
				return false;
			}
			if (obj is HotKey)
			{
				return obj.GetHashCode() == this.GetHashCode();
			}
			else
			{
				return false;
			}
		}

		public static bool operator ==(HotKey a, HotKey b)
			=> a.GetHashCode() == b.GetHashCode();

		public static bool operator !=(HotKey a, HotKey b)
			=> a.GetHashCode() != b.GetHashCode();

		public override string ToString()
		{
			StringBuilder sb = new StringBuilder();
			if (Modifiers.HasFlag(KeyModifiers.MOD_ALT))
			{
				sb.Append("Alt+");
			}
			if (Modifiers.HasFlag(KeyModifiers.MOD_CONTROL))
			{
				sb.Append("Ctrl+");
			}
			if (Modifiers.HasFlag(KeyModifiers.MOD_SHIFT))
			{
				sb.Append("Shift+");
			}
			if (Modifiers.HasFlag(KeyModifiers.MOD_WIN))
			{
				sb.Append("Win+");
			}
			sb.Append(Enum.GetName(typeof(Keys), Key) ?? ((int)Key).ToString());
			return sb.ToString();
		}
	}

	struct ScreenColorEffect
	{
		public float[,] Matrix { get; }
		public string Description { get; }

		public ScreenColorEffect(float[,] matrix, string description)
			: this()
		{
			Matrix = matrix;
			Description = description;
		}

	}

	enum ConfigurationLocation
	{
		Embedded,
		AppData,
		WorkingDirectory,
	}
}
